src/zh/cartoon18/src/eu/kanade/tachiyomi/extension/zh/cartoon18/Cartoon18.kt

```diff
@@ -0,0 +1,230 @@
+            title = document.selectFirst("div.content h1.title")!!.ownText()
+            thumbnail_url = document.selectFirst("div.content h1.title ~ div.row a img")!!.attr("src")
+            genre = genres.text()
+            if (authors.size > 1) {
+                author = authors[1].text().replace(",(\\S)".toRegex(), ", $1")
+            }
+            if (descriptions.size > 1) {
+                description = descriptions[1].text()
+            }
+        }
+    }
+
+    override fun chapterListParse(response: Response): List<SChapter> {
+        val document = response.asJsoup()
+        val chapters = document.select("div.content h1.title + div a")
+        return chapters.map {
+            SChapter.create().apply {
+                url = it.attr("href")
+                name = it.text()
+            }
+        }.reversed()
+    }
+
+    override fun pageListParse(response: Response): List<Page> {
+        val document = response.asJsoup()
+        val images = document.select("div#lightgallery a img")
+        return images.mapIndexed { index, image ->
+            Page(index, imageUrl = image.attr("src"))
+        }
+    }
+
+    override fun imageUrlParse(response: Response) = throw UnsupportedOperationException()
+
+    override fun getFilterList(): FilterList {
+        launchIO { fetchKeywords() }
+        return FilterList(
+            SortFilter(),
+            if (keywordsList.isEmpty()) {
+                Filter.Header("Tap 'Reset' to load keywords")
+            } else {
+                KeywordFilter(keywordsList)
+            },
+        )
+    }
+
+    private open class QueryFilter(
+        name: String,
+        values: Array<String>,
+        private val queryName: String,
+        private val queryValues: Array<String>,
+        state: Int = 0,
+    ) : Filter.Select<String>(name, values, state) {
+        fun addQueryTo(builder: HttpUrl.Builder) =
+            builder.addQueryParameter(queryName, queryValues[state])
+    }
+
+    private class SortFilter : QueryFilter(
+        "Sort by",
+        arrayOf("Latest", "Popular", "Recommended", "Best"),
+        "sort",
+        arrayOf("created", "hits", "score", "likes"),
+        state = 2,
+    )
+
+    class Keyword(val name: String, val value: String)
+
+    private var keywordsList: List<Keyword> = emptyList()
+
+    private class KeywordFilter(keywords: List<Keyword>) : QueryFilter(
+        "Keyword",
+        keywords.map { it.name }.toTypedArray(),
+        "q",
+        keywords.map { it.value }.toTypedArray(),
+    )
+
+    /**
+     * Inner variable to control how much tries the keywords request was called.
+     */
+    private var fetchKeywordsAttempts: Int = 0
+
+    /**
+     * Fetch the keywords from the source to be used in the filters.
+     */
+    private fun fetchKeywords() {
+        if (fetchKeywordsAttempts < 3 && keywordsList.isEmpty()) {
+            try {
+                keywordsList = client.newCall(keywordsRequest()).execute()
```

Comment:
merge the request and parse fucntions into this function. Madara does it this way so sources may override those specific methods, which isn't the case here.